src/backend/joanie/tests/core/test_api_order.py

```diff
@@ -1912,3 +1970,356 @@ def test_api_order_create_no_limit(self):
+        """
+        user = factories.UserFactory()
+        token = self.generate_token_from_user(user)
+        course = factories.CourseFactory()
+        product = factories.ProductFactory(courses=[course], price=0.00)
+        organization = product.course_relations.first().organizations.first()
+
+        data = {
+            "course": course.code,
+            "organization": str(organization.id),
+            "product": str(product.id),
+        }
+        response = self.client.post(
+            "/api/v1.0/orders/",
+            data=data,
+            content_type="application/json",
+            HTTP_AUTHORIZATION=f"Bearer {token}",
+        )
+        self.assertEqual(response.status_code, 201)
+        self.assertEqual(response.json()["state"], enums.ORDER_STATE_DRAFT)
+        order = models.Order.objects.get(id=response.json()["id"])
+        response = self.client.patch(
+            f"/api/v1.0/orders/{order.id}/submit/",
+            HTTP_AUTHORIZATION=f"Bearer {token}",
+        )
+        self.assertEqual(response.status_code, 201)
+        order.refresh_from_db()
+        self.assertEqual(order.state, enums.ORDER_STATE_VALIDATED)
+
+    def test_api_order_create_no_billing_address_to_validation(self):
+        """
+        Create an order on a fee product should be done in 3 steps.
+        First create the order in draft state. Then submit the order by
+        providing a billing address should pass the order state to `submitted`
+        and return payment information. Once the payment has been done, the order
+        should be validated.
+        """
+        user = factories.UserFactory()
+        token = self.generate_token_from_user(user)
+        course = factories.CourseFactory()
+        product = factories.ProductFactory(courses=[course])
+        organization = product.course_relations.first().organizations.first()
+
+        data = {
+            "course": course.code,
+            "organization": str(organization.id),
+            "product": str(product.id),
+        }
+
+        response = self.client.post(
+            "/api/v1.0/orders/",
+            data=data,
+            content_type="application/json",
+            HTTP_AUTHORIZATION=f"Bearer {token}",
+        )
+        self.assertEqual(response.status_code, 201)
+        self.assertEqual(response.json()["state"], enums.ORDER_STATE_DRAFT)
+        order_id = response.json()["id"]
+        billing_address = BillingAddressDictFactory()
+        data["billing_address"] = billing_address
+        response = self.client.patch(
+            f"/api/v1.0/orders/{order_id}/submit/",
+            data=data,
+            content_type="application/json",
+            HTTP_AUTHORIZATION=f"Bearer {token}",
+        )
+        self.assertEqual(response.status_code, 201)
+        order = models.Order.objects.get(id=order_id)
+        self.assertEqual(order.state, enums.ORDER_STATE_SUBMITTED)
+
+        InvoiceFactory(order=order)
+        order.validate()
```

Comment:
Just add an `order.refresh_from_db()` after validate just to ensure that the state update has been persisted.